### PR TITLE
Enhance logging

### PR DESCRIPTION
**Pull request by luckyxiaoqiang**
### Describe what this PR does / why we need it
- Logging supports auto conversion of error interface.
- Refine logging param when it is an error.
- Add more tests.

### Special notes for reviews

### CODE REVIEW

File: logging/logging.go [hunk 1]
```diff
@@ -126,6 +126,15 @@ func caller(depth int) (file string, line int) {
 	return
 }
 
+// toSafeJSONString converts to valid JSON string, as the original string may contain '\\', '\n', '\r', '\t' and so on.
+func toSafeJSONString(s string) []byte {
+	if data, err := json.Marshal(s); err == nil {
+		return data
+	} else {
+		return []byte(`"JSON Marshal Error"`)
```
Comment (louyuting) [inline logging/logging.go:134]: Maybe we just return `"` + s + `"`  when occurs error?

File: logging/logging.go [hunk 2]
```diff
@@ -189,23 +197,27 @@ func AssembleMsg(depth int, logLevel, msg string, err error, keysAndValues ...in
 				kStr = fmt.Sprintf("%+v", k)
 			}
 			sb.WriteByte(',')
-			sb.WriteByte('"')
-			sb.WriteString(kStr)
-			sb.WriteByte('"')
+			data := toSafeJSONString(kStr)
+			sb.Write(data)
 			sb.WriteByte(':')
 			vStr, vIsStr := v.(string)
```
Comment (louyuting) [inline logging/logging.go:203]: Maybe we could refine code here:
```go
                   switch v.(type) {
			case string:
				data := toSafeJSONString(v.(string))
				sb.Write(data)
			case error:
				data := toSafeJSONString(v.(error).Error())
				sb.Write(data)
			default:
				if vbs, err := json.Marshal(v); err != nil {
					s := fmt.Sprintf("%+v", v)
					data := toSafeJSONString(s)
					sb.Write(data)
				} else {
					sb.Write(vbs)
				}
			}
```

Comment (luckyxiaoqiang) [reply logging/logging.go]: Good advice. Addressed.

File: logging/logging.go (hunk 1 above)
Comment (luckyxiaoqiang) [reply logging/logging.go]: Agree. Addressed.

**Review by louyuting — APPROVED** [review]
nice work and looking forward more

**Pull request merged**